apps/blockchain/lib/blockchain/contract/create_contract.ex

```diff
@@ -122,30 +122,32 @@ defmodule Blockchain.Contract.CreateContract do
     state_after_init = exec_env.account_interface.state
 
     contract_creation_cost = creation_cost(output)
+    insufficient_gas = remaining_gas < contract_creation_cost
 
-    insufficient_gas_before_homestead =
-      remaining_gas < contract_creation_cost and Header.is_before_homestead?(params.block_header)
-
-    resultant_gas =
-      cond do
-        state_after_init == nil -> 0
-        insufficient_gas_before_homestead -> remaining_gas
-        true -> remaining_gas - contract_creation_cost
-      end
-
-    resultant_state =
-      cond do
-        state_after_init == nil ->
-          params.state
-
-        insufficient_gas_before_homestead ->
-          state_after_init
-
-        true ->
-          Account.put_code(state_after_init, address, output)
-      end
-
-    {resultant_state, resultant_gas, accrued_sub_state}
+    if insufficient_gas && EVM.Configuration.fail_contract_creation?(params.config) do
+      {:error, {params.state, 0, SubState.empty()}}
+    else
+      resultant_gas =
+        cond do
+          state_after_init == nil -> 0
+          insufficient_gas -> remaining_gas
+          true -> remaining_gas - contract_creation_cost
+        end
+
+      resultant_state =
+        cond do
```

Comment:
This `cond` and the one above are kind of confusing to me. Or at least, I can't quite figure out why we are checking those conditions. I know you didn't really change this and it's mostly indentation, but do you know why we check these conditions? And is it something we could encapsulate in a well-named function?